### ISSUE TITLE
[class.cdtor] does not specify the case when the delete-expression is invoked during the construction and destruction

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#307]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[basic.life] p6 says
> For an object under construction or destruction, see [class.cdtor]. Otherwise, such a pointer refers to allocated storage ([basic.stc.dynamic.allocation]), and using the pointer as if the pointer were of type void* is well-defined. Indirection through such a pointer is permitted but the resulting lvalue may only be used in limited ways, as described below.
The program has undefined behavior if: 
>> - the pointer is used as the operand of a delete-expression,
>>- the pointer is used to access a non-static data member or call a non-static member function of the object, or
>> - [...]

Consider this example:
````cpp
struct A{
  A():a((delete this,0)){}
  int a;
  std::string b;
  ~A(){}
};
int main(){
  A* ptr = new A;
}
````
In this example, we invoke delete-expression when the object is under construction, however,  [class.cdtor] does not make a wording to rule the case. Obviously, the bullet in the list does not apply to the case when the object is under construction, because we can access the non-static data member when the object is under construction or destruction, otherwise, it would violate the second bullet if the bullet applies to the case. Moreover, the rule explicitly says the cases when the object is under construction or destruction is ruled by [class.cdtor] and the other cases will be regulated by the common rules. 

### Suggested resolution
[class.cdtor] should tell the complete story, it should say how the case would be if we invoke delete-expression when the object is under construction or destruction. Or, [class.cdtor] should explicitly say the other cases are well-defined if they are not covered in the following paragraphs in the subclause.

**Comment by languagelawyer** [comment]
Also, nothing about placement-new into `this`, and lotsa other things.

I think
> [class.cdtor] does not make a wording to rule the case

is usually considered «implicit UB».

**Comment by xmh0511** [comment]
> Also, nothing about placement-new into `this`, and lotsa other things.
> 
> I think
> 
> > [class.cdtor] does not make a wording to rule the case
> 
> is usually considered «implicit UB».

I think we have large discussions on this part in other issues, the document should explicitly say what case is UB and what is not. Instead of establishing the concept on an unwritten statement: if the subclause says nothing, it implies UB.

**Comment by frederick-vs-ja** [comment]
I think [[class.dtor] p18](https://eel.is/c++draft/class.dtor#18) has already implied that `delete this` during destruction is generally UB, although the status is not very clear during construction.

**Comment by frederick-vs-ja** [comment]
Delete-expressions are a bit complicated. IMO we should divide this into two issues:
- Whether destroying `*this` during construction should be UB?
- Any attempt to complete the initialization of an object whose storage has been released should be UB, which doesn't seem clarified.

**Comment by xmh0511** [comment]
> Whether destroying *this during construction should be UB?

In general, `delete this`(in the construction) will invoke the destructor that may access the non-static data member, at this time, if the non-static data member has not been initialized(i.e. whose lifetime has not begun), the behavior is UB. SO, in terms of this respect, should we consider `delete this` during construction to always be harmful?

**Comment by frederick-vs-ja** [comment]
It seems to me that deallocation and storage resuing should be handled together (#341).

The remain issue (to me) is that what if `delete this` uses a destroying `operator delete`, but that `operator delete` doesn't deallocate the storage or not even destroy the object?

**Comment by t3nsor** [comment]
> I think [[class.dtor] p18](https://eel.is/c++draft/class.dtor#18) has already implied that `delete this` during destruction is generally UB, although the status is not very clear during construction.

Yes, I think if we add "or has not begun" to the end of [class.dtor]/18, that's enough for this particular case.

For the non-class case, perhaps some similar change is needed to [expr.call]/4, e.g., add to the end of the last sentence "; the behavior is undefined if that object's lifetime has ended or has not begun".

Is it time to make this into a core issue yet?

**Comment by xmh0511** [comment]
> > I think [[class.dtor] p18](https://eel.is/c++draft/class.dtor#18) has already implied that `delete this` during destruction is generally UB, although the status is not very clear during construction.
> 
> Yes, I think if we add "or has not begun" to the end of [class.dtor]/18, that's enough for this particular case.
> 
> For the non-class case, perhaps some similar change is needed to [expr.call]/4, e.g., add to the end of the last sentence "; the behavior is undefined if that object's lifetime has ended or has not begun".
> 
> Is it time to make this into a core issue yet?

I still think we should move these rules into [class.cdtor] subclause because [basic.life] p6 and p7 have already said:
> For an object under construction or destruction, see [class.cdtor].  

This means all complete stories regarding objects of class type that are under construction or destruction is regulated by [class.cdtor].

>

**Comment by jensmaurer** [comment]
One could argue that the destructor call has started, thus the lifetime has ended per https://eel.is/c++draft/basic.life#1.4

And thus, calling delete is undefined behavior per https://eel.is/c++draft/basic.life#6.1.

And that means https://eel.is/c++draft/class.dtor#18 is entirely redundant and should be turned into a note.

**Comment by xmh0511** [comment]
However,  the destruction and construction of class type are not regulated by https://eel.is/c++draft/basic.life#6.1. 

> For an object under construction or destruction, see [class.cdtor]. **Otherwise,** ...

**Comment by t3nsor** [comment]
> I still think we should move these rules into [class.cdtor] subclause because [basic.life] p6 and p7 have already said:
> 
> > For an object under construction or destruction, see [class.cdtor].
> 
> This means all complete stories regarding objects of class type that are under construction or destruction is regulated by [class.cdtor].
> 

I think if something is UB regardless of whether it occurs before construction or during construction, then it doesn't need to be in [class.cdtor]. Likewise, when something is UB regardless of whether it occurs during destruction or after destruction. Obviously, in cases where the rules are more relaxed during construction than before construction, those rules need to be documented in [class.cdtor].

**Comment by jensmaurer** [comment]
@t3nsor, it seems the only rule called out in https://eel.is/c++draft/basic.life#6 that isn't modified or overridden by [class.cdtor] is the "delete" case. Given the existing wording ("Otherwise", it makes sense to duplicate the delete rule in [class.cdtor], I think.

**Comment by jensmaurer** [comment]
CWG2757

**Comment by languagelawyer** [comment]
Hm, no exception for reusing by an object nested within? I think one may want to use an `std::byte` array as a storage during construction, like some `std::optional` constructos do

**Comment by jensmaurer** [comment]
I was hoping the cross-reference after "reuse" would make it clear that "nested within" is not a reuse.

**Comment by languagelawyer** [comment]
Ehm, ok, what does
> [is reused by an object that is not nested within _o_](https://timsong-cpp.github.io/cppwp/n4868/basic.life#1.5)

mean? That nested within is not a reuse (and thus does not end the lifetime), or that reuse which is also nested within does not end the lifetime?

**Comment by jensmaurer** [comment]
I'd like to point out that [basic.life] p6 also talks about (undecorated) reuse, as does p5, with the understanding that "reuse that is nested within" is not harmful to the parent object.
